contrib/babelfishpg_tsql/src/linked_servers.c

```diff
@@ -0,0 +1,1024 @@
+
+			void *val[MAX_COLS_SELECT];
+
+			if (erc == FAIL)
+			{
+				ereport(ERROR,
+					(errcode(ERRCODE_FDW_UNABLE_TO_CREATE_EXECUTION),
+						errmsg("Failed to get results from query %s", query)
+					));
+			}
+
+			/* store the column metadata if present */
+			colcount = LINKED_SERVER_NUM_COLS(lsproc);
+
+			for(i = 0; i < colcount; i++)
+			{
+				/* Let us process column metadata first */
+				coltype[i] = LINKED_SERVER_COL_TYPE(lsproc, i + 1);
+				colname[i] = LINKED_SERVER_COL_NAME(lsproc, i + 1);
+				collen[i] = LINKED_SERVER_COL_LEN(lsproc, i + 1);
+			}
+
+			if(colcount > 0)
+			{
+				/* check to see if caller supports us returning a tuplestore */
+				if (rsinfo == NULL || !IsA(rsinfo, ReturnSetInfo))
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("set-valued function called in context that cannot accept a set")));
+
+				if (!(rsinfo->allowedModes & SFRM_Materialize))
+					ereport(ERROR,
+							(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+							errmsg("materialize mode required, but it is not allowed in this context")));
+
+				/* Build tupdesc for result tuples. */
+				tupdesc = CreateTemplateTupleDesc(colcount);
+
+				/* 
+				 * If column name is NULL, we set the column name as "?column?" (default 
+				 * column name set by PG if column name is NULL). We have logic in the
+				 * babelfishpg_tds extension to set the column name length 0 on the wire
+				 * if we come across this column name.
+				 */
+				for(i = 0; i < colcount; i++)
+				{
+					Oid tdsTypeOid = tdsTypeToOid(coltype[i]);
+
+					/* 
+					 * Current TDS client library has a limitation where it can send
+					 * column types like nvarchar as varchar in column metadata, so
+					 * check with out previously computed tuple descriptor to see what
+					 * should be the actual data type. At the moment there is no other way.
+					 */
+					if ((tdsTypeOid == VARCHAROID) || (tdsTypeOid == TEXTOID) || (common_utility_plugin_ptr && ((*common_utility_plugin_ptr->lookup_tsql_datatype_oid)("binary"))))
+					{
+						Form_pg_attribute att = TupleDescAttr(rsinfo->expectedDesc, (AttrNumber) i);
+						tdsTypeOid = att->atttypid;
+					}
+
+					TupleDescInitEntry(tupdesc, (AttrNumber) (i + 1), colname[i], tdsTypeOid, tdsTypeLen(coltype[i], collen[i], false), 0);
+				}
+				tupdesc = BlessTupleDesc(tupdesc);
+
+				per_query_ctx = rsinfo->econtext->ecxt_per_query_memory;
+				oldcontext = MemoryContextSwitchTo(per_query_ctx);
+
+				tupstore = tuplestore_begin_heap(true, false, work_mem);
+				rsinfo->returnMode = SFRM_Materialize;
+				rsinfo->setResult = tupstore;
+				rsinfo->setDesc = tupdesc;
+
+				MemoryContextSwitchTo(oldcontext);
+
+				/* fetch the rows */
+				while ((erc = LINKED_SERVER_NEXT_ROW(lsproc)) != NO_MORE_ROWS)
+				{
+					/* for each row */
+					Datum	*values = palloc0(sizeof(SIZEOF_DATUM) * colcount);
+					bool	*nulls = palloc0(sizeof(bool) * colcount);
+
+					MemSet(nulls, false, sizeof(nulls));
```

Comment:
not required, 0 is already interpreted as false

Reply:
removed MemSet